File: algorithms/M/Minimize the Difference Between Target and Chosen Elements/Minimize the Difference Between Target and Chosen Elements.js
var minimizeTheDifference = function(mat, target) {
    const MAX = Number.MAX_SAFE_INTEGER;
    
    const m = mat.length;
    const n = mat[0].length;
    
    const memo = [];
    
    for (let i = 0; i < m; ++i) {
        memo[i] = new Array(4901).fill(MAX);
    }
    
    return topDown(0, 0);
    
    function topDown(row, sum) {
        if (row === m) return Math.abs(target - sum);
        if (memo[row][sum] != MAX) return memo[row][sum];
        
        let min = MAX;
        
        mat[row].sort((a, b) => a - b); 
        
        const set = new Set(mat[row]);
       
        for (const num of set) {
           const res = topDown(row + 1, sum + num);
            
           if (res > min) break;
           min = res;
        }
        
        memo[row][sum] = min;
        
        return min;
    }
};

File: scripts/algorithms/M/Minimize the Difference Between Target and Chosen Elements/Minimize the Difference Between Target and Chosen Elements.cpp
// Runtime: 891 ms (Top 82.1%) | Memory: 15.48 MB (Top 61.3%)

class Solution 
{
public:
    int dp[8000][71];
    int n,m;
    int find(vector<vector<int>>&mat,int r,int sum,int &target)
    {
        if(r>=n)
        {
            return abs(sum-target);
        }
        if(dp[sum][r]!=-1)
        {
            return dp[sum][r];
        }
        int ans=INT_MAX;
        for(int i=0;i<m;i++)
        {
            ans=min(ans,find(mat,r+1,sum+mat[r][i],target));
            if(ans==0)
            {
                break;
            }
        }
        return dp[sum][r]=ans;
    }
    int minimizeTheDifference(vector<vector<int>>& mat, int target) 
    {
        memset(dp,-1,sizeof(dp));
        n=mat.size();
        m=mat[0].size();
        return find(mat,0,0,target);
    }
};